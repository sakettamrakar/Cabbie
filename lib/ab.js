import crypto from 'crypto';
export function titleVariantForRoute(origin, destination) {
    const list = (process.env.EXP_TITLE_ROUTES || '').split(',').map(s => s.trim().toLowerCase()).filter(Boolean);
    const key = `${origin}-${destination}`.toLowerCase();
    if (list.includes(key))
        return 'priceFirst';
    const h = crypto.createHash('md5').update(key).digest('hex');
    const n = parseInt(h.slice(0, 4), 16);
    const pct = n / 0xffff;
    return pct < 0.1 ? 'priceFirst' : 'benefitFirst';
}
export function buildFareTitle(origin, destination, baseFare, variant, brand) {
    if (variant === 'priceFirst' && baseFare) {
        return `${origin} to ${destination} Taxi Fare from â‚¹${baseFare} | ${brand}`;
    }
    return `${origin} to ${destination} Taxi Fare | ${brand}`;
}
