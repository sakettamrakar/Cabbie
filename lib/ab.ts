import crypto from 'crypto';
import type { TitleVariant } from './seo';

export function titleVariantForRoute(origin:string,destination:string): TitleVariant {
  const list = (process.env.EXP_TITLE_ROUTES||'').split(',').map(s=>s.trim().toLowerCase()).filter(Boolean);
  const key = `${origin}-${destination}`.toLowerCase();
  if(list.includes(key)) return 'priceFirst';
  const h = crypto.createHash('md5').update(key).digest('hex');
  const n = parseInt(h.slice(0,4),16);
  const pct = n / 0xffff;
  return pct < 0.1 ? 'priceFirst' : 'benefitFirst';
}

export function buildFareTitle(origin:string,destination:string,baseFare:number|undefined|null, variant:TitleVariant, brand:string){
  if(variant==='priceFirst' && baseFare){
    return `${origin} to ${destination} Taxi Fare from â‚¹${baseFare} | ${brand}`;
  }
  return `${origin} to ${destination} Taxi Fare | ${brand}`;
}