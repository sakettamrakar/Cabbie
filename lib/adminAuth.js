import jwt from 'jsonwebtoken';
import { getCsrfToken, validateCsrf } from './csrf';
const JWT_SECRET = process.env.ADMIN_JWT_SECRET || 'dev_admin_secret_change_me';
export function verifyJwtFromRequest(req) {
    const raw = req.cookies['admin_session'];
    if (!raw)
        return null;
    try {
        const decoded = jwt.verify(raw, JWT_SECRET);
        return { userId: decoded.sub, email: decoded.email };
    }
    catch {
        return null;
    }
}
export function requireAdminAuth(req, res) {
    const ctx = verifyJwtFromRequest(req);
    if (!ctx) {
        res.status(401).json({ ok: false, error: 'unauthorized' });
        return null;
    }
    return ctx;
}
export function requireCsrf(req, res) {
    if (['GET', 'HEAD', 'OPTIONS'].includes(req.method || ''))
        return true;
    const token = req.headers['x-csrf-token'] || (req.body && req.body._csrf);
    if (!token || Array.isArray(token)) {
        res.status(403).json({ ok: false, error: 'missing_csrf' });
        return false;
    }
    const cookieToken = req.cookies['csrf_token'];
    if (!cookieToken) {
        res.status(403).json({ ok: false, error: 'missing_csrf_cookie' });
        return false;
    }
    if (!validateCsrf(String(token), cookieToken)) {
        res.status(403).json({ ok: false, error: 'invalid_csrf' });
        return false;
    }
    return true;
}
export function issueCsrf(res) {
    const { cookieValue, token } = getCsrfToken();
    // 2h expiry
    res.setHeader('Set-Cookie', `csrf_token=${cookieValue}; Path=/; Max-Age=${60 * 120}; SameSite=Lax`);
    return token;
}
