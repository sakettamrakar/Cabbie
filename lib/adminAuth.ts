import jwt from 'jsonwebtoken';
import type { NextApiRequest, NextApiResponse } from 'next';
import { getCsrfToken, validateCsrf } from './csrf';

const JWT_SECRET = process.env.ADMIN_JWT_SECRET || 'dev_admin_secret_change_me';

export interface AdminContext { userId:number; email:string; }

export function verifyJwtFromRequest(req:NextApiRequest): AdminContext | null {
  const raw = req.cookies['admin_session'];
  if(!raw) return null;
  try {
    const decoded:any = jwt.verify(raw, JWT_SECRET);
    return { userId: decoded.sub, email: decoded.email };
  } catch { return null; }
}

export function requireAdminAuth(req:NextApiRequest,res:NextApiResponse): AdminContext | null {
  const ctx = verifyJwtFromRequest(req);
  if(!ctx){ res.status(401).json({ ok:false, error:'unauthorized' }); return null; }
  return ctx;
}

export function requireCsrf(req:NextApiRequest,res:NextApiResponse): boolean {
  if(['GET','HEAD','OPTIONS'].includes(req.method||'')) return true;
  const token = req.headers['x-csrf-token'] || (req.body && req.body._csrf);
  if(!token || Array.isArray(token)){ res.status(403).json({ ok:false, error:'missing_csrf' }); return false; }
  const cookieToken = req.cookies['csrf_token'];
  if(!cookieToken){ res.status(403).json({ ok:false, error:'missing_csrf_cookie' }); return false; }
  if(!validateCsrf(String(token), cookieToken)){ res.status(403).json({ ok:false, error:'invalid_csrf' }); return false; }
  return true;
}

export function issueCsrf(res:NextApiResponse){
  const { cookieValue, token } = getCsrfToken();
  // 2h expiry
  res.setHeader('Set-Cookie', `csrf_token=${cookieValue}; Path=/; Max-Age=${60*120}; SameSite=Lax`);
  return token;
}
