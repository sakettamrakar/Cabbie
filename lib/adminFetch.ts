export async function adminFetch(input:RequestInfo | URL, init:RequestInit = {}){
  const csrf = typeof window !== 'undefined' ? sessionStorage.getItem('csrfToken') : null;
  const method = (init.method || 'GET').toUpperCase();
  const headers = new Headers(init.headers || {});
  if(method !== 'GET' && method !== 'HEAD'){
    if(csrf && !headers.has('x-csrf-token')) headers.set('x-csrf-token', csrf);
    if(!headers.has('Content-Type') && !(init.body instanceof FormData)) headers.set('Content-Type','application/json');
  }
  return fetch(input, { ...init, headers });
}
