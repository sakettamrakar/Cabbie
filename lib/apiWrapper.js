import { newCorrelationId } from './ids';
import { makeError, sendError } from './errors';
export function withApi(handler, opts = {}) {
    return async function wrapped(req, res) {
        const cid = newCorrelationId();
        req.correlation_id = cid;
        res.correlation_id = cid;
        try {
            if (opts.cors) {
                res.setHeader('Vary', 'Origin');
                const origin = req.headers.origin || '';
                const allow = process.env.NODE_ENV !== 'production' && /localhost|127\.0\.0\.1/.test(String(origin));
                if (allow) {
                    res.setHeader('Access-Control-Allow-Origin', String(origin));
                    res.setHeader('Access-Control-Allow-Credentials', 'true');
                    res.setHeader('Access-Control-Allow-Headers', 'Content-Type, x-csrf-token');
                    res.setHeader('Access-Control-Allow-Methods', 'GET,POST,PUT,DELETE,OPTIONS');
                }
                if (req.method === 'OPTIONS') {
                    res.status(204).end();
                    return;
                }
            }
            await Promise.resolve(handler(req, res));
        }
        catch (err) {
            if (!err.code)
                err = makeError('INTERNAL_ERROR', 500, err.message || 'Internal error');
            sendError(res, err);
        }
    };
}
