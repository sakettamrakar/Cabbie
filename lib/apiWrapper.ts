import type { NextApiHandler, NextApiRequest, NextApiResponse } from 'next';
import { newCorrelationId } from './ids';
import { makeError, sendError } from './errors';

interface WrapperOpts { cors?: boolean }

export function withApi(handler:NextApiHandler, opts:WrapperOpts = {}): NextApiHandler {
  return async function wrapped(req:NextApiRequest,res:NextApiResponse){
    const cid = newCorrelationId();
    (req as any).correlation_id = cid;
    (res as any).correlation_id = cid;
    try {
      if(opts.cors){
        res.setHeader('Vary','Origin');
        const origin = req.headers.origin || '';
        const allow = process.env.NODE_ENV !== 'production' && /localhost|127\.0\.0\.1/.test(String(origin));
        if(allow){
          res.setHeader('Access-Control-Allow-Origin', String(origin));
          res.setHeader('Access-Control-Allow-Credentials','true');
          res.setHeader('Access-Control-Allow-Headers','Content-Type, x-csrf-token');
          res.setHeader('Access-Control-Allow-Methods','GET,POST,PUT,DELETE,OPTIONS');
        }
        if(req.method==='OPTIONS'){ res.status(204).end(); return; }
      }
      await Promise.resolve(handler(req,res));
    } catch(err:any){
      if(!err.code) err = makeError('INTERNAL_ERROR',500, err.message || 'Internal error');
      sendError(res, err);
    }
  };
}
