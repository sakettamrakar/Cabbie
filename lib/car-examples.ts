const FALLBACK_CAR_EXAMPLES: Record<string, readonly string[]> = {
  economy: ['Maruti Swift', 'Maruti WagonR'],
  comfort: ['Maruti Swift Dzire', 'Toyota Etios', 'Hyundai Aura'],
  'premium-7-seater': ['Maruti Ertiga'],
  'luxury-7-seater': ['Toyota Innova'],
};

const normalizeCategory = (category?: string) => {
  if (!category) {
    return '';
  }

  return category
    .toLowerCase()
    .normalize('NFKD')
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/(^-|-$)/g, '');
};

export const getFallbackCarExamples = (category?: string): string[] => {
  const key = normalizeCategory(category);
  if (!key) {
    return [];
  }

  const entries = FALLBACK_CAR_EXAMPLES[key];
  return entries ? [...entries] : [];
};

export const resolveCarExamples = (
  cab: { carExamples?: string[] | null; category?: string } | null | undefined,
): string[] => {
  if (!cab) {
    return [];
  }

  if (cab.carExamples && cab.carExamples.length > 0) {
    return cab.carExamples;
  }

  return getFallbackCarExamples(cab.category);
};

export const __TESTING__ = {
  FALLBACK_CAR_EXAMPLES,
  normalizeCategory,
};
