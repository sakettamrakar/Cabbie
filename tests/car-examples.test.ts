import { getFallbackCarExamples, resolveCarExamples, __TESTING__ } from '../lib/car-examples';

describe('car example fallbacks', () => {
  it('returns predefined entries for normalized categories', () => {
    expect(getFallbackCarExamples('Economy')).toEqual(['Maruti Swift', 'Maruti WagonR']);
    expect(getFallbackCarExamples('Comfort')).toEqual([
      'Maruti Swift Dzire',
      'Toyota Etios',
      'Hyundai Aura',
    ]);
    expect(getFallbackCarExamples('Premium (7-Seater)')).toEqual(['Maruti Ertiga']);
    expect(getFallbackCarExamples('Luxury — 7 seater')).toEqual(['Toyota Innova']);
  });

  it('returns empty list when nothing matches', () => {
    expect(getFallbackCarExamples('Unknown Category')).toEqual([]);
    expect(getFallbackCarExamples(undefined)).toEqual([]);
  });

  it('prefers cab provided examples', () => {
    const cab = {
      category: 'Economy',
      carExamples: ['Existing Example'],
    };

    expect(resolveCarExamples(cab)).toEqual(['Existing Example']);
  });

  it('falls back to mapping when cab examples missing', () => {
    const cab = { category: 'Premium – 7 Seater', carExamples: [] };

    expect(resolveCarExamples(cab)).toEqual(['Maruti Ertiga']);
  });

  it('exposes mapping for snapshot guardrails', () => {
    expect(__TESTING__.FALLBACK_CAR_EXAMPLES).toMatchSnapshot();
  });
});
