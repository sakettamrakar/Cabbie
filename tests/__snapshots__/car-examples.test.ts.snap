// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`car example fallbacks exposes mapping for snapshot guardrails 1`] = `
{
  "comfort": [
    "Maruti Swift Dzire",
    "Toyota Etios",
    "Hyundai Aura",
  ],
  "economy": [
    "Maruti Swift",
    "Maruti WagonR",
  ],
  "luxury-7-seater": [
    "Toyota Innova",
  ],
  "premium-7-seater": [
    "Maruti Ertiga",
  ],
}
`;
